### ISSUE TITLE
Generating HTML

### ISSUE THREAD

**Issue by wiseoldman95** [bel2125/civetweb#60]
I can implement an interface which would make the following code possible:

```
int helloWorld(struct mg_connection *conn, void *cbdata)
{
	struct mg_request_info * req_info = mg_get_request_info(conn);	
	char output[1000];
			
	
	el html=c(NULL,"html"); //create a new element called html
		el head=c(html,"head");
			el meta=c(head,"meta");
			sa(meta,"charset","utf-8"); //set attribute
		el body=c(html,"body");
			t(body,"Hello world!"); //text node
			c(body,"br");
		el table=c(body,"table");
		sa(table,"border","1");
		el td,tr;
			td=c(table,"td");
					tr=c(td,"tr");
						t(tr,"A third cell!");
					tr=c(td,"tr");
						t(tr,"A fourth cell!");
			td=c(table,"td");
					tr=c(td,"tr");
						t(tr,"A cell!");
					tr=c(td,"tr");
						t(tr,"A second cell!");
			
	int size=xmlToString(html,output,1000);
	
	 mg_printf(conn,
	"HTTP/1.1 200 OK\r\n"
	"Content-Type: text/html\r\n"
	"Content-Length: %d\r\n"
	"\r\n"
	"%s",size, output);
	
	xmlDelete(html);
	return 0;
}
```

This would generate the following HTML:
```
<html><head><meta charset="utf-8" /></head>
<body>Hello world!<br /><table>...etc...</table></body>
```

Are you interested in this?

**Comment by bel2125** [comment]
Would this work for C or C++? The above example is almost C but not strict C (MSVC would not accept `el body` after `sa(...)` in C).

I guess it would it be possible to write XHTML/XML too?

Do you need a maximum size like 10000 for `xmlToString(html,output,1000);` or would it be possible to let the function handle the memory allocation?

You know that and agree that every code you post here or add to civetweb has to have an MIT license? Of course you would get credits in the contributors list (https://github.com/bel2125/civetweb/blob/master/CREDITS.md)

I could imagine this as an additional header file, defining the functions c, t, sa, ... 
Probably there is even a way to encapsulate the sequence ...

    int size=xmlToString(html,output,1000);

     mg_printf(conn,
    "HTTP/1.1 200 OK\r\n"
    "Content-Type: text/html\r\n"
    "Content-Length: %d\r\n"
    "\r\n"
    "%s",size, output);

    xmlDelete(html);

... into a single mg_send_XML call.

**Comment by wiseoldman95** [comment]
Both the auto allocation part and the mg_send_XML are great ideas. The mg_send_XML would also function as the de-allocator.

The example code I provided is valid C compiled with gcc with no extra flags at all. I noticed it does throw some errors with the -pedantic flag. I will use the strictest form of C89 when implementing, just in case.

I am aware of the MIT license.

**Comment by wiseoldman95** [comment]
From an efficiency perspective, pre-allocating a buffer on the stack(e.g. creating a 10k char array) makes a lot more sense than malloc.

Perhaps the class should do that?
e.g. in xml_header.c
```
#define XML_BUFFER_SIZE 10000
char buffer[XML_BUFFER_SIZE];
```

This raises a related subquestion: Could Civetweb serve two pages concurrently via threading? If so, this option might not  be viable (Two functions writing to the same buffer at the same time is a mess).

**Comment by bel2125** [comment]
Civetweb is multi threaded, but every response is generated by only one thread.
The output buffer and the elements will for one response will only be accessed from one thread.

Putting 10000 bytes on the stack will generate warnings in some code analysis tools.
A fully featured system like a desktop PC running Windows or Linux will not have any problem with 10k or 100k on the stack. For embedded hardware this is more critical.
I think it is better to allocate it dynamically with mg_malloc/mg_free. This is handled effectively by the C library of the operating system. If performance optimization in this area is required, one could replace the mg_malloc/mg_free functions by some private memory pool or whatever. This would optimize all memory allocations, not just the one for the HTML response generated here.

**Comment by wiseoldman95** [comment]
Perhaps we should leave the choice of allocaiton to the user?

**Comment by wiseoldman95** [comment]
I have a working implementation here:

https://github.com/wiseoldman95/FeatherXML

**Comment by bel2125** [comment]
> Perhaps we should leave the choice of allocation to the user?

Lua uses a default allocator, and allows the user to overload it: 
http://www.lua.org/manual/5.2/manual.html#lua_newstate
Would this be an option?

Maybe these functions could be made available for the Lua server pages and Lua scripts as well. Currently there is LuaXML to do that, but this does not offer a C api.
Do you plan to do a XML parser as well?

**Comment by wiseoldman95** [comment]
I have zero experience with Lua and I don't quite understand how it fits in with c code. Could you explain the Lua part further?

I do not plan to do an XML parser any time soon, sorry.

Currently, the toString function expects to receive an output buffer (a char *), it doesn't care where that buffer came from, the user has the choice of using the heap or stack. 
See an example code here: https://github.com/bel2125/civetweb/pull/61/files#diff-0

**Comment by bel2125** [comment]
Civetweb allows to use Lua script pages to generate the web contents, instead of using C to generate the contents. C code can be easily made accessible for Lua scripts, depending on the interface of the library in C.
How do you know how large the buffer is inside the implementation?
How do you know how large the buffer has to be outside the implementation?

**Comment by wiseoldman95** [comment]
```
How do you know how large the buffer is inside the implementation?
How do you know how large the buffer has to be outside the implementation?
```
I am not sure I understand what you mean.
The implementation does not need a buffer at all except when calling `int fxml_ToString(el elem, char * buffer)`

In such a case, the output is saved at `char * buffer`.

**Comment by bel2125** [comment]
I mean, in your test
https://github.com/wiseoldman95/FeatherXML/blob/master/main.c
you use a buffer `char output[5000];` in line 5.

Why 5000? What if you use `char output[15];` instead?
try

    int size;
    int a;
    char output[15];
    int b;

    a = 0x12345678; b=0x9ABCDEF0;

    ...

    print("a=%x, b=%x\r\n", a, b);
    size=fxml_ToString(html,output)+15; /*+15 is for <!DOCTYPE html>*/
    print("a=%x, b=%x\r\n", a, b);

    ...

**Comment by wiseoldman95** [comment]
My example will output about about 300 bytes of html, it won't fit in 15
bytes.

**Comment by bel2125** [comment]
This is what I mean.
How do you know that in advance?
E.g., imagine you creating a html table from a list (of files, images, users, whatever) where you do not know how many elements this will be. How can you know how large the `output` buffer has to be when you call `? 
If you call it with small buffer, `fxml_ToString` may even overwrite data or cause a crash. That's why `sprintf` is deprecated nowadays, and why civetweb uses `mg_snprintf`. This has an extra argument with the size of the buffer, and snprinf will never write more than this buffer size.

**Comment by wiseoldman95** [comment]
I see. This is a matter of tradeoff between memory and CPU:
Either allocate a really big buffer you know you'll never fill, or come up with a way to calculate the needed buffer size on runtime. It is possible to give  the user the choice.

Here's my proposed fix:
Suppose we have this function: `int fxml_bufferSize(el)` which calculates the needed.

Someone with lots of Ram could stick to the way I did it in the example, trading memory for CPU.

Someone with a limited stack could do this:
```
int size;
char * output;
int b;

...xml stuff...

output=malloc(fxml_bufferSize(html));
size=fxml_ToString(html,output)+15;
free(output);
...
```

**Comment by bel2125** [comment]
What do you think about this:

    int required_size;
    int available_size;
    int actual_size;

    ...xml stuff...

    /* dynamic case */
    char * output;
    required_size = fxml_docSize(html) + 1;
    output = malloc(required_size);
    if (output) {available_size = required_size;}
    else { ... handle error }

    /* alternatively: use stack */
    char output[1000];
    available_size = sizeof(output);
    
    /* common */
    actual_size=fxml_ToString(html,output, available_size);
    if (actual_size < 0) {
        /* error code -1 is "buffer too small" */
        handle error
        return
    }

    printf( ... size+15 ... <!DOCTYPE html> ... output );
    free(output); /* only in dynamic case */

**Comment by wiseoldman95** [comment]
Yes, this is better. I definitely agree with `fxml_ToString` having a max buffer size as a parameter.
Do you have any more changes in mind?

**Comment by wiseoldman95** [comment]
Hmm, although this is functional, I think it's complicating things. We could let go of the stack, and then it would be that simple:

```
...xml stuff...
int size;
char * output=fxml_toString(html,&size);
if (size==-1)
{
  /*Handle error*/
}
printf( ... size+15 ... <!DOCTYPE html> ... output );
free(output);
fxml_delete(html);
```
fxml_toString would call fxml_docSize internally.

**Comment by wiseoldman95** [comment]
I have a working implementation of this:
```
int main()
{
	el html,head,body,table,td,tr,meta;
	
	int size;
	char * output;
	
	html=c(NULL,"html");
		head=c(html,"head");
			meta=c(head,"meta");
			sa(meta,"charset","utf-8");
		body=c(html,"body");
			t(body,"Hello world!");
			c(body,"br");
		table=c(body,"table");
		sa(table,"border","1");
			tr=c(table,"tr");
					td=c(tr,"td");
						t(td,"Cell 1");
					td=c(tr,"td");
						t(td,"Cell 2");
			tr=c(table,"tr");
					td=c(tr,"td");
						t(td,"Cell 3");
					td=c(tr,"td");
						t(td,"Cell 4");
			

	fxml_toString(html,&output,&size); 
	if (size==-1)
	{
		printf("Malloc failed!\n");
		return -1;
	}
	size+=15; /*+15 is for <!DOCTYPE html>*/
	printf(
	"HTTP/1.1 200 OK\r\n"
	"Content-Type: text/html\r\n"
	"Content-Length: %d\r\n"
	"\r\n"
	"<!DOCTYPE html>%s",size, output);
	
	fxml_Delete(html);
	mg_free(output); /*If you're not using Civetweb, this can be 
	replaced with "free" */
	return 0;
}
```

**Comment by wiseoldman95** [comment]
For anyone interested, the pull request is here: https://github.com/bel2125/civetweb/pull/61

**Comment by bel2125** [comment]
Unfortunately, I haven't had time to have a look at this topic for the last three weeks.
Now it seems the link above (https://github.com/wiseoldman95/FeatherXML) does not work anymore.

Still, in an ideal case civetweb would need an XML writer and parser for both, C and Lua.

**Comment by wiseoldman95** [comment]
Yes, I removed the repository because I felt the code is not production-ready yet. I intend to re-publish it later.

It started as a minimalist XML/HTML generator for my own personal experiments with Civetweb and c/c++. I found it an order of magnitude easier than outputting plain html with printf. 

So, it is a replacement for printf and not a fully-fledged xml engine. In other words, it lacks parsing and is only good at outputting things. I do not intend to implement parsing, because my sole purpose is replacing printf. If anyone else adds parsing later, I wouldn't mind it.

This might be a little naive but: Are there any practical uses for parsing in an embedded c/c++ server? The only use I can think of is fetching a remote HTML, parsing and manipulating it, then sending the modified version to users.

**Comment by bel2125** [comment]
> Are there any practical uses for parsing in an embedded c/c++ server?

You could fetch a request with XML data, e.g. a SOAP request (http://en.wikipedia.org/wiki/SOAP#Example_message).

If you can choose the protocol on both sides, client and server, you may use other protocols, e.g. JSON. However, XML is a standard and every major modern higher language (e.g., C#, Java, ...) offers facilities to read and write it, so XML is often used to change data between applications.

**Comment by bel2125** [comment]
Also Civetweb is not only an embedded webserver, but you can use it as a standalone web server. You can use Lua scripts (and standard CGI) to generate dynamic contents.
This is why I would like to have an XML parser/writer for both, C and Lua.

**Comment by wiseoldman95** [comment]
One alternative option is to incorporate MiniXML with Civetweb (Just like you did with sqlite), it is more fully fledged than my XML engine, and it includes parsing.
http://www.msweet.org/projects.php?Z3

**Comment by bel2125** [comment]
MiniXML has a problematic license (LGPL).
We can only accept MIT licensed code.

**Comment by kainjow** [comment]
IMO this feature shouldn't be part of the library. Let civetweb stay as a simple web server and let clients choose which HTML writing libraries they want to use. For example, [Mustache](http://mustache.github.io) is a far cleaner way of generating HTML.

**Comment by sunsetbrew** [comment]
I am with Kevin on this.  All of the solutions are too limited and there
are better ones which can be added on by the users.

**Comment by bel2125** [comment]
CivetWeb has three different types of source code:

1) The web server core, written in plain C, not an application by itself but exporting a low level C API. The basis functionality - all others are built on top of it.
This is basically civetweb.c + civetweb.h + *.inl

2) The standalone web server with Lua extension. This is the executable from SourceForge you can use to build a full web application. This is 1) plus the third_party directory.

3) Some convenience layers for embedding, e.g. 1) plus CivetServer.cpp - 
This is just another, higher level interface around the functions from 1)

Any XML parser/writer will certainly not be in 1), but I need an XML parser for the standalone interpreter 2). If the same parser can be used for 3), with a "USE_..." define as well, it's even better.
You will always be able to just use 1) and your own XML library or whatever you want.
Some of the past and current issues are actually requests for a somewhat higher level interface than the core interface of 1). In principle one could also split it in three different components and handle it in different repositories, but I think this does not make maintenance any easier.
The MIT license is required for 1), 2) and 3), credits and release notes are also not separated.


Anyway, the original repository above does not exist anymore, and MiniXML can not be used.
So for the moment I would just keep LuaXML as a parser for 2) and everybody has to care about an XML parser/writer on his own.

**Comment by wiseoldman95** [comment]
>Anyway, the original repository above does not exist anymore

I still have the code. I took it down because I felt it's too dirty to be public. I've cleaned it up since then, but didn't have the time to thoroughly test it. Note that it was only an XML generator, not a parser.

I can set up a repo if you're interested, but I probably won't do any further work on it.

**Comment by gbjbaanb** [comment]
You could always try tinyxml2 (zlib licence, which is practically the same words as MIT). I've used it commercially and its simple enough, it was also designed to be small and suitable for old Android devices so (I assume) suitable for many embedded systems.

http://www.grinninglizard.com/tinyxml2/index.html

**Comment by bel2125** [comment]
Thank you for the information.

I saw tinyxml2 is C++, while all other components (CivetWeb, Lua, SQLite, ...) are C.
CivetWeb runs on a lot of different platforms - for some of them adding C++ code is not that easy, and I don't want to support the C++ build process for all of them - dealing with the plain C build on all those platforms and different compilers and compiler versions is difficult enough.

Everyone is free to add C++ code on his own, in his distribution, but I don't want to do this as a required component in the plain build.

**Comment by gbjbaanb** [comment]
No worries, I don't even think about the differences between C and C++ any more, but maybe I don't have the restrictions of many embedded compilers!

**Comment by LogicParrot** [comment]
I'll look search for the code. If found, I'll put it in a repo hoping someone might continue where I left off.
